docs/adrs/002-use-rolling-update-with-eviction.md

```diff
@@ -0,0 +1,82 @@
+---
+status: accepted
+date: 2024-05-17
+deciders: Kamil Kołodziej, Radosław Szuma, Jakub Pisarek
+informed: Fishjam Team, Cloud Team
+---
+# Change deployment strategy to eliminate downtime
+
+## Context and Problem Statement
+
+In the current situation when we do a deployment we are shuting down the server and every room with that.
+This is an awful user experience. Another issue is the fact that any ongoing recordings are lost.
+
+## Decision Drivers
+
+* No downtime while deployment happens
+* 0 lost recordings during deployment
+
+## Considered Options
+
+* Rolling update with eviction
+* Active room migration
+
+## Decision Outcome
+
+Chosen option: "Rolling update with eviction", because
+while second option sounds great, we acknowledge that solution is not trivial and we want to fix downtimes ASAP. First solution is a good starting point to remove existing problems and allow us to move forward.
+
+### Consequences
+
+Every new deployment will need to trigger fishjam process which will handle the shutdown.
+What that effectively means is that we are leaving the responsibility of triggering deployment to external orchestration tool (what is usually the case) but we handle the process inside the app.
+
+External process will trigger fishjam shutdown process by a SIGTERM.
+This is gonna mark fishjam as one that no longer accepts new rooms.
+Once every room is closed and all of the recordings are computed we are going to shutdown that instance of fishjam.
+
+This is applicable one by one for every instance in cluster, although it may take some time to deployment a new version and we may have 2 different versions on the cluster at the same time, we are accepting that tradeoff.
```

Comment:
As we discussed, having two different versions in a cluster might be a problem as those versions might not be backward and forward compatible. We can take a look what rabbit does https://www.rabbitmq.com/docs/upgrade

Reply:
Based on the Rabbit approach they allow rolling updates in some scenarios but when there is some breaking change (version incompatibility) they require you to do full stop or blue-green deployment. 

Their approach to blue-green deployment is to completely create a second cluster and slowly move traffic to the new one while draining the queues from the old one. This is very interesting approach but I don't find it necessary to dig it into now. I would propose moving forward with rolling updates approach and full stop. If that's gonna happen often then I suggest we come back to this discussion.